### PR TITLE
[BugFix] Sort key column prohibit NULL.

### LINKED ISSUE #13772: [pk_sort_key]The result is wrong when sort key is nonsequenced/ not_pk/ all column
### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
```
CREATE TABLE `primary_table` (
                `k1` date NOT NULL,
                `k2` datetime NOT NULL,
                `k3` varchar(20) NOT NULL,
                `k4` varchar(20) NOT NULL,
                `k5` boolean NOT NULL,
                `v1` tinyint NULL,
                `v2` smallint NULL,
                `v3` int NULL,
                `v4` bigint NULL,
                `v5` largeint NULL,
                `v6` float NULL,
                `v7` double NULL,
                `v8` decimal(27,9) NULL
            ) ENGINE=OLAP
            PRIMARY KEY(`k1`, `k2`, `k3`, `k4`, `k5`)
            COMMENT "OLAP"
            DISTRIBUTED BY HASH(`k1`, `k2`, `k3`) BUCKETS 3
            ORDER BY(`v5`, `v6`, `v7`, `v8`, `v1`, `v2`, `v3`, `v4`, `k1`, `k2`, `k3`, `k4`, `k5`)
            PROPERTIES (
                "replication_num" = "3",
                "enable_persistent_index" = "true",
                "storage_format" = "v2"
            );
CREATE TABLE `primary_table` (
                `k1` date NOT NULL,
                `k2` datetime NOT NULL,
                `k3` varchar(20) NOT NULL,
                `k4` varchar(20) NOT NULL,
                `k5` boolean NOT NULL,
                `v1` tinyint NULL,
                `v2` smallint NULL,
                `v3` int NULL,
                `v4` bigint NULL,
                `v5` largeint NULL,
                `v6` float NULL,
                `v7` double NULL,
                `v8` decimal(27,9) NULL
            ) ENGINE=OLAP
            PRIMARY KEY(`k1`, `k2`, `k3`, `k4`, `k5`)
            COMMENT "OLAP"
            DISTRIBUTED BY HASH(`k1`, `k2`, `k3`) BUCKETS 3
            ORDER BY(`v5`, `k4`, `k2`, `v3`, `v1`)
            PROPERTIES (
                "replication_num" = "3",
                "enable_persistent_index" = "true",
                "storage_format" = "v2"
            );
CREATE TABLE `primary_table` (
                `k1` date NOT NULL,
                `k2` datetime NOT NULL,
                `k3` varchar(20) NOT NULL,
                `k4` varchar(20) NOT NULL,
                `k5` boolean NOT NULL,
                `v1` tinyint NULL,
                `v2` smallint NULL,
                `v3` int NULL,
                `v4` bigint NULL,
                `v5` largeint NULL,
                `v6` float NULL,
                `v7` double NULL,
                `v8` decimal(27,9) NULL
            ) ENGINE=OLAP
            PRIMARY KEY(`k1`, `k2`, `k3`, `k4`, `k5`)
            COMMENT "OLAP"
            DISTRIBUTED BY HASH(`k1`, `k2`, `k3`) BUCKETS 3
            ORDER BY(`v5`, `v3`, `v1`)
            PROPERTIES (
                "replication_num" = "3",
                "enable_persistent_index" = "true",
                "storage_format" = "v2"
            );
```
2. INSERT INTO '....'
```
curl --location-trusted -u root: -T /home/disk4/lvchenyang/codebak/StarRocksTest/lib/../common/data/basic_types_data_value_null -XPUT -H label:stream_load_1669028711220 -H columns:__op='upsert' http://ip:port/api/test_pk_sort_key_5da78d74_698c_11ed_81f5_00163e0e550b/primary_table/_stream_load
```
3. SELECT  '....'
```
select sum(cast(v1 as int)), sum(v2), sum(v3), max(v4), max(v5), min(v6), min(v7), sum(v8) from primary_table where k1 = '2020-06-23' and k2 <= '2020-06-23 18:11:00';
```
### Expected behavior (Required)
(-512, -33030656, -2199022731776, -9223372036854774785, '-18446744073709550593', -3.1, -3.14, Decimal('-2692.608'))
### Real behavior (Required)
(575, -21672897, -1443108664257, -9223372036854774785, '-18446744073709550593', -3.1, -3.14, Decimal('-1763.553000000'))
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
+----------------------------+
| current_version()          |
+----------------------------+
| BRANCH-2.5-RELEASE 7d36d91 |
+----------------------------+
```

### PR DESCRIPTION
**Pull request by Linkerist**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #13772

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java [hunk 1]
```diff
@@ -119,12 +119,21 @@ public static void analyze(CreateTableStmt statement, ConnectContext context) {
...
             if (keysDesc == null || keysDesc.getKeysType() != KeysType.PRIMARY_KEYS) {
                 throw new IllegalArgumentException("only primary key support sort key");
             }
+            List<String> columnNames = columnDefs.stream().map(ColumnDef::getName).collect(Collectors.toList());
+            for (String sortKey : statement.getSortKeys()) {
+                int idx = columnNames.indexOf(sortKey);
+                ColumnDef cd = columnDefs.get(idx);
+                cd.setPrimaryKeyNonNullable();
+                if (cd.isAllowNull()) {
+                    throw new SemanticException("primary key column[" + cd.getName() + "] cannot be nullable");
```
Comment (decster) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java:133]: sort key column

Comment (Linkerist) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java]: changed.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 8 / 9 (88.89%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|8|9|88.89%|[133]|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java [hunk 2]
```diff
@@ -119,12 +119,21 @@ public static void analyze(CreateTableStmt statement, ConnectContext context) {
         statement.setCharsetName(analyzeCharsetName(statement.getCharsetName()).toLowerCase());
 
         KeysDesc keysDesc = statement.getKeysDesc();
+        List<ColumnDef> columnDefs = statement.getColumnDefs();
         if (statement.getSortKeys() != null) {
             if (keysDesc == null || keysDesc.getKeysType() != KeysType.PRIMARY_KEYS) {
                 throw new IllegalArgumentException("only primary key support sort key");
             }
+            List<String> columnNames = columnDefs.stream().map(ColumnDef::getName).collect(Collectors.toList());
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java:127]: Add a description of the problem and solution in commit message.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java [hunk 3]
```diff
@@ -119,12 +119,21 @@ public static void analyze(CreateTableStmt statement, ConnectContext context) {
         statement.setCharsetName(analyzeCharsetName(statement.getCharsetName()).toLowerCase());
 
         KeysDesc keysDesc = statement.getKeysDesc();
+        List<ColumnDef> columnDefs = statement.getColumnDefs();
         if (statement.getSortKeys() != null) {
             if (keysDesc == null || keysDesc.getKeysType() != KeysType.PRIMARY_KEYS) {
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java:124]: If you have already guaranteed the sort key is a subset of primarykey, under which instance, the sort key can be nullable?

Comment (chaoyli) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java:124]: If not, why sort key must be non-nullable?

**Review by chaoyli — APPROVED** [review]

**Comment by sevev** [comment]
I think this pr(https://github.com/StarRocks/starrocks/pull/15641) can resolve the issue and this pr can be closed

**Review by decster — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Linkerist** [comment]
> I think this pr(#15641) can resolve the issue and this pr can be closed

yeah, i mentioned this former PR in the latter one.
image

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request closed without merging**